morango/api/viewsets.py

```diff
@@ -338,83 +370,80 @@ def create(self, request):  # noqa: C901
+        response_status = status.HTTP_201_CREATED
+        if result != transfer_status.COMPLETED:
+            response_status = status.HTTP_202_ACCEPTED
 
-        transfersession.server_fsic = json.dumps(
-            DatabaseMaxCounter.calculate_filter_max_counters(requested_filter)
+        return response.Response(
+            self.get_serializer(controller.context.transfer_session).data,
+            status=response_status,
         )
-        transfersession.save()
 
-        if not is_a_push:
+    def update(self, request, *args, **kwargs):
+        if not kwargs.get("partial", False):
+            return response.Response(
+                "Only PATCH updates allowed", status=status.HTTP_405_METHOD_NOT_ALLOWED
+            )
 
-            # queue records to get ready for pulling
-            with OperationLogger("Queueing records into buffer", "Queueing complete"):
-                _queue_into_buffer(transfersession)
-            # update records_total on transfer session object
-            records_total = Buffer.objects.filter(
-                transfer_session=transfersession
-            ).count()
-            transfersession.records_total = records_total
-            transfersession.save()
+        update_stage = request.data.pop("transfer_stage", None)
+        if update_stage is not None:
+            # if client is trying to update `transfer_stage`, then we use the controller to proceed
+            # to the stage, but wait for completion if both do not support async
+            controller = SessionController.build_local(
+                request=request, transfer_session=self.get_object(), enable_logging=True
+            )
+            if self.async_allowed():
+                controller.proceed_to(update_stage)
+            else:
+                controller.proceed_to_and_wait_for(update_stage)
 
-        return response.Response(
-            serializers.TransferSessionSerializer(transfersession).data,
-            status=status.HTTP_201_CREATED,
-        )
+        return super(TransferSessionViewSet, self).update(request, *args, **kwargs)
 
-    def perform_destroy(self, transfersession):
-        if transfersession.push:
-            # if no records were transferred, we can safely skip the next steps
-            if (
-                transfersession.records_total is not None
-                and transfersession.records_total > 0
-            ):
-                # dequeue into store and then delete records
-                with OperationLogger(
-                    "Dequeuing records into store", "Dequeuing complete"
-                ):
-                    _dequeue_into_store(transfersession)
-                # update database max counters but use latest fsics on server
-                DatabaseMaxCounter.update_fsics(
-                    json.loads(transfersession.client_fsic),
-                    certificates.Filter(transfersession.filter),
-                )
+    def list(self, request, *args, **kwargs):
```

Comment:
e.g. here, we could avoid having to do this by just not including the ListModelMixin in the viewset in the first place.

https://www.django-rest-framework.org/api-guide/viewsets/#example_3